src/main/java/game/data/chunk/version/Chunk_1_18.java

```diff
@@ -69,66 +62,70 @@ protected void parse(DataTypeProvider dataProvider) {
+           }
+
+           // For some reason, there's a chance that the packet has no more data. Instead of erroring,
+           // just check if the packet has any more data.
+           if (dataProvider.hasNext()) {
+               // parse blocks
+               section.setBlocks(dataProvider.readLongArray(dataProvider.readVarInt()));
+
+               // biomes
+               if (dataProvider.hasNext()) {
+                   section.setBiomePalette(Palette.readPalette(dataProvider));
+                   section.setBiomes(dataProvider.readLongArray(dataProvider.readVarInt()));
+               }
+           }
+
+           // May replace an existing section or a null one
+           setChunkSection(sectionY, section);
+
+           // servers don't (always?) include containers in the list of block_entities. We need to know that these block
+           // entities exist, otherwise we'll end up not writing block entity data for them
+           if (containsBlockEntities(blockPalette)) {
+               findBlockEntities(section, sectionY);
+           }
+       }
+    }
 
-            // For some reason, there's a chance that the packet has no more data. Instead of erroring,
-            // just check if the packet has any more data.
-            if(dataProvider.hasNext()) {
-                // parse blocks
-                section.setBlocks(dataProvider.readLongArray(dataProvider.readVarInt()));
+    private void findBlockEntities(ChunkSection section, int sectionY) {
+        BlockEntityRegistry blockEntities = RegistryManager.getInstance().getBlockEntityRegistry();
+        GlobalPalette globalPalette = GlobalPaletteProvider.getGlobalPalette(getDataVersion());
 
-                // biomes
-                if(dataProvider.hasNext()) {
-                    section.setBiomePalette(Palette.readPalette(dataProvider));
-                    section.setBiomes(dataProvider.readLongArray(dataProvider.readVarInt()));
-                }
-            }
+        for (int y = 0; y < 16; y++) {
+            for (int z = 0; z < 16; z++) {
+                for (int x = 0; x < 16; x++) {
+                    BlockState state = globalPalette.getState(section.getNumericBlockStateAt(x, y, z));
 
-            // May replace an existing section or a null one
-            setChunkSection(sectionY, section);
-
-            // For some reason that I can't figure out, servers don't include containers
-            // in the list of block_entities. We need to know that these block entities
-            // exist, otherwise we'll end up not writing block entity data which results
-            // in certain block entities that don't render!
-            for(SpecificTag tag : blockPalette.toNbt()) {
-                // Only iterate over the chunk section if the palette contains a block entity
-                String blockName = tag.get("Name").stringValue();
-                if(isBlockEntity(blockName)) {

```

Comment:
The reason I had this additional iteration and if condition here is that I wanted to only iterate over the chunk section if there was a valid block entity in that chunk section - this prevents unnecessarily iterating over every block in the chunk section if there isn't one (e.g. don't iterate over a chunk section which consists of only air)